scanpy/_utils/__init__.py

```diff
@@ -560,6 +560,156 @@ def _elem_mul_dask(x: DaskArray, y: DaskArray) -> DaskArray:
+    column_divide = axis == 1
+    # In this case `data` matches the order given by `nonzero` and we can scale directly.
+    if (row_divide and dividend_type == sparse.csr_matrix) or (
+        column_divide and dividend_type == sparse.csc_matrix
+    ):
+        divisor_indexed = np.ravel(divisor)[
+            dividend.nonzero()[types.index(dividend_type)]
+        ]
+        scaled_data = dividend.data / divisor_indexed
+        return dividend_type(
+            (scaled_data, dividend.indices, dividend.indptr), shape=dividend.shape
+        )
+    # We thus need to do the type conversion so that `data` and `nonzero` match by converting to the other for division and then back to return.
+    other_sparse_type = list(set(types).difference({dividend_type}))[0]
+    return dividend_type(divide(other_sparse_type(dividend), divisor, axis))
+
+
+@divide.register(np.ndarray)
+def _(
+    dividend: np.ndarray,
+    divisor: np.ndarray,
+    axis: Literal[0, 1],
+    *,
+    out: np.ndarray | None = None,
+) -> np.ndarray:
+    dividend = np.divide(dividend, divisor, out=out)
+    return dividend
+
+
+@divide.register(DaskArray)
+def _(
+    dividend: DaskArray,
+    divisor: np.ndarray | DaskArray,
+    axis: Literal[0, 1],
+    *,
+    out: None = None,
+) -> DaskArray:
+    if out is not None:
+        raise TypeError(
+            "`out` is not `None`. Do not do in-place modifications on dask arrays."
+        )
+
+    import dask.array as da
+
+    row_divide = axis == 0
+    column_divide = axis == 1
+
+    if isinstance(divisor, DaskArray):
+        warnings.warn("Rechunking divisor in user operation", UserWarning)
+        if row_divide and not dividend.chunksize[0] == divisor.chunksize[0]:
+            divisor = divisor.rechunk((dividend.chunksize[0], 1))
+        elif column_divide and (
+            (
+                len(divisor.chunksize) == 1
+                and dividend.chunksize[1] != divisor.chunksize[0]
+            )
+            or (
+                len(divisor.chunksize) == 2
+                and dividend.chunksize[1] != divisor.chunksize[1]
+            )
+        ):
+            divisor = divisor.rechunk((1, dividend.chunksize[1]))
+    elif row_divide:
+        divisor = da.from_array(divisor, chunks=(dividend.chunksize[axis], 1))
+    elif column_divide:
+        divisor = da.from_array(divisor, chunks=(1, dividend.chunksize[axis]))
+
+    return da.map_blocks(divide, dividend, divisor, axis, meta=dividend._meta, out=out)
+
+
+@overload
+def axis_sum(
+    X: sparse.spmatrix,
+    *,
+    axis: Optional[Union[tuple[Literal[0, 1], ...], Literal[0, 1]]] = None,  # noqa: UP007
+    dtype: Optional[np.typing.DTypeLike] = None,  # noqa: UP007
+) -> np.matrix:
+    ...
+
+
+@singledispatch
+def axis_sum(
+    X: np.ndarray,
+    *,
+    axis: Optional[Union[tuple[Literal[0, 1], ...], Literal[0, 1]]] = None,  # noqa: UP007
+    dtype: Optional[np.typing.DTypeLike] = None,  # noqa: UP007
+) -> np.ndarray:
```

Comment:
Please remove those `noqa` things. You e.g. do `@singledispatch(np.ndarray)`, that’s less ugly.

Reply:
Sorry @flying-sheep this has to do with an `|` operator.  What does `@singledispatch(np.ndarray)` have to with that?   Also can you call `singledispatch` directly on something like that?

Reply:
I assumed you blocked reformatting these lines because that interferes with singledispatch somehow.

Since singledispatch supports an alternative form of use, we should use that instead of making our type hints uglier.

>  Also can you call singledispatch directly on something like that?

Yup, of course! I’m not trying to mislead you by adding broken code examples :wink:

Reply:
I am sorry if I am misunderstanding but

```python
from functools import singledispatch
@singledispatch(int)
def foo(a: int):
     return a
```
raises an error.  I also still don't quite see how this affects the `noqa`. The `noqa` comment out came from just before https://github.com/scverse/scanpy/pull/2856/commits/48a1a1e3c9412efb7d435a1f57a42e4e70aef649.  I was a little confused about why that was happening since the `|` operator works elsewhere with no complaints.  I think it had to do with the multiple generics in the union.

Reply:
I believe this will get rid of the need for the noqa

```suggestion
@singledispatch
def axis_sum(
    X: np.ndarray,
    *,
    axis: tuple[Literal[0, 1], ...] | Literal[0, 1] | None = None,
    dtype: np.typing.DTypeLike | None = None,
) -> np.ndarray:
```

-----


@flying-sheep, `@singledispatch` is what registers the generic. You can't pass a type as argument. I believe it also shouldn't care about the annotations, that's just `.register`. That should also only effect the first argument.

Reply:
@ivirshup I will try again, but I believe that is what I tried previously.  It's possible I erroneously altered the `dtype` one but it was an issue with python 3.9 only and I think the code I have might be necessary.  I will look again though

Reply:
Ok yes https://dev.azure.com/scverse/scanpy/_build/results?buildId=5927&view=logs&j=5ea502cf-d418-510c-3b5f-c4ba606ae534&t=534778bb-2f86-5739-7d3c-59518f7b5a2b&l=31 there is definitely something bizarre going on!  But somehow calling `axis_sum.register(DaskArray)` fixes it?  Ok!